StRoot/StFcsWaveformFitMaker/PeakAnaPainter.h

```diff
@@ -0,0 +1,106 @@
+/*
+Author: David Kapukchyan
+@[June 12, 2022]
+> Painter now sets the found peak line color
+
+@[May 10, 2022]
+> Changed how options work. Graph options now supported. Use semicolon separated list for graph, peak and stat options respectively. Added style options for the various lines. Moved lines for the peak windows to the 'PeakWindow' class.
+
+@[March 23, 2022]
+> Found a bug where calling "Draw" on ROOT objects is bad because it adds the TLine objects created by this class to the pad. This means the pad can delete those objects which causes seg fault errors. The proper method to call is "Paint". This is because "Draw" for most ROOT objects only adds objects to the TPad's list so that later the pad can call the "Paint" method of the added class. This means when "Draw" is called on a *PeakAna* object it gets added to the pad so that "Paint" can be called. The "Paint" method will now draw all the lines needed. However, if "Draw" is called the lines are also added to the TPad's list. This is not needed since it is sufficient that only the *PeakAna* class be in the list so that the lines are not deleted by TPad.
+
+ If that paint method involves a "Draw" call then it only adds objects to the pad without calling their paint method
+
+@[March 14, 2022]
+> Separated the peak stats box painting method and added options for it
+
+@[March 9, 2022]
+> This class will be used to draw PeakAna. It is a first try after reading ROOT code and and seeing that by having a separate painter class is better due to how TCanvas objects work and drawing in ROOT in general works. Also after splitting up the various classes I realized that it really is easier to have a separate painter and analysis class. The painter class essentially assumes a canvas/pad exists and only draws on it the things it needs independent of what else is on the canvas/pad. This is the philosophy to keep in mind when writing painter classes.
+
+Painter class for PeakAna
+*/
+
+#ifndef STROOT_STFCSWAVEFORMFITMAKER_PEAKANAPAINTER_H_
+#define STROOT_STFCSWAVEFORMFITMAKER_PEAKANAPAINTER_H_
+
+//C++ Headers
+#include <vector>
+#include <utility> //std::move (Needs C++ 11) and std::pair
+
+//ROOT Headers
+#include "TObject.h"
+#include "TKey.h"
+#include "TH1.h"
+
+//Custom Headers
+#include "PeakWindow.h"
+#include "PeakAnaVirtualPainter.h"
+#include "PeakAna.h"
+
+class PeakAnaPainter : public PeakAnaVirtualPainter
```

Comment:
Besides this class PeakAnaPainter, is there another class that inherits from PeakAnaVirtualPainter? If not, are you sure you need to create the hierarchy with virtual inheritance for this simple case?

Reply:
This follows the ROOT TH1 code design for TVirtualHistPainter and THistPainter (among others), and it is necessary for the following reason. If you have just a PeakAnaPainter object then PeakAna must have a pointer to PeakAnaPainter which you can forward declare. However, forward declaration of a PeakAnaPainter is not sufficient since you need to call methods of PeakAnaPainter in PeakAna. This means that you must include PeakAnaPainter and you can't just declare it. This creates another problem because you also need to forward declare PeakAna in PeakAnaPainter because PeakAnaPainter has a pointer to PeakAna. Now you have the same problem as before because forward declaring PeakAna is not enough since you call on PeakAna methods in PeakAnaPainter. This means that both PeakAna must include PeakAnaPainter, and PeakAnaPainter must include PeakAna which is not possible. The virtual painter solves this problem because I can include the PeakAnaVirtualPainter which doesn't require the existence of PeakAna. Sorry for the long winded and maybe confusing explanation but I did spend a lot of time figuring this out and realized that without a virtual interface, a separate analysis and painter class could not exist.

Reply:
Forward declaration of PeakAna should work in this header file since you refer to it via a pointer only. You can do #include "PeakAna.h" in the cxx file. Similarly, in PeakAna.h you can forward declare PeakAnaPainter and include in the implementation file.

Early versions of ROOT is not the best place to pick up design patterns 😄

Reply:
Thanks, that worked! I thought I had tried something similar to that but it didn't work that's why I opted to try to use the virtual methods I saw in the ROOT code.